Parsers/ASimNetworkSession/Parsers/ASimNetworkSessionCiscoASA.yml

```diff
@@ -0,0 +1,369 @@
+                                        ProtocolFromLog endswith "41", "41",
+                                        ""),
+                        NetworkRuleName = trim(@'"',NetworkRuleName),
+                        DeviceAction = trim_start(@"\s*",Action),
+                        SimplifiedDeviceAction = trim_start(@"\s*",Action)
+                | project-away ProtocolFromLog, Action, Destination*, Source*;
+      let all_302013_302015_parsed = allLogs
+                | where DeviceEventClassID in ("302013","302015") and isnotempty(SourceIP)
+                | parse Message with * ":  " * " " * " " * " connection " NetworkSessionId " for " SrcInterfaceName ":" * "/" * " (" SrcNatIpAddr "/" SrcNatPortNumber:int ")" SrcUsername "to " DstInterfaceName ":" * "/" * " (" DstNatIpAddr "/" DstNatPortNumber:int ")" DstUsername
+                | extend SrcUsername = trim(@"[()]", SrcUsername),
+                        DstUsername = trim(@"[()]", DstUsername)
+                | extend 
+                        SrcUsernameType = iif(isnotempty(SrcUsername),"Windows",""),
+                        DstUsernameType = iif(isnotempty(DstUsername),"Windows",""),
+                        SimplifiedDeviceAction = DeviceAction,
+                        SessionId = NetworkSessionId;
+      let all_302013_302015_unparsed = allLogs
+                | where DeviceEventClassID in ("302013","302015") and isempty(SourceIP)
+                | parse Message with * ":  " DeviceAction " " NetworkDirection " " Protocol " connection " NetworkSessionId " for " SrcInterfaceName ":" SrcIpAddr "/" SrcPortNumber:int " (" SrcNatIpAddr "/" SrcNatPortNumber:int ")" SrcUsername "to " DstInterfaceName ":" DstIpAddr "/" DstPortNumber:int " (" DstNatIpAddr "/" DstNatPortNumber:int ")" DstUsername
+                | extend SrcUsername = trim(@"[()]", SrcUsername),
+                        DstUsername = trim(@"[()]", DstUsername),
+                        NetworkDirection = case(NetworkDirection == "inbound", "Inbound",
+                                                    NetworkDirection == "outbound", "Outbound",
+                                                    "")
+                | extend 
+                        SrcUsernameType = iif(isnotempty(SrcUsername),"Windows",""),
+                        DstUsernameType = iif(isnotempty(DstUsername),"Windows",""),
+                        SimplifiedDeviceAction = DeviceAction,
+                        SessionId = NetworkSessionId;
+      let all_302014_302016 = allLogs
+                | where DeviceEventClassID in ("302014", "302016")
+                | parse kind=regex Message with * ": " DeviceAction " " Protocol " connection " NetworkSessionId " for " SrcInterfaceName ":" SrcIpAddr "/" SrcPortNumber:int @"[(]?" SrcUsername @"[)]? to " DstInterfaceName ":" DstIpAddr "/" DstPortNumber:int @"[(]?" DstUsername1 @"[)]? [Dd]uration " NetworkDuration " bytes " DstBytes:long @"\s?" Remaining 
```

Comment:
Few regex comments:
- @"[(]?" -> no need for a character group here, you really want \(?.
- Don't change now, but for next time:
  - you can break the parse to the part that does not need kind-regex and do just the rest with a second parse with kind=regex.
  - The parse statement can really be without kind=regex: 
    - trim out the () from user names after the parse. 
    - remove the extra spaces from Remining in the extract_all statement
    - replace the [dD] with a *. I am not sure it would work, if not, the only option is to replace duration with Duration in a previous extend.

Reply:
Resolved point 1, noted the remaining.